### PR TITLE
Internationalization of kiwix-serve's backend 

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Addresses the C++ part of kiwix/kiwix-tools#59

This PR introduces facilities for supporting internationalized textual output from the kiwix-serve backend (C++ code).

- Message templates (in mustache format) are stored in JSON files under `static/i18n/` (one file per language). Those JSON files are compiled into C++ code by a new script `kiwix-compile-i18n`. 
- Access to messages from C++ code is through the `kiwix::ParameterizedMessage` class. `ParameterizedMessage(msgId, [msgParams])` creates a message object with the specified id and (optional) parameters. The text of the message for the desired language must be obtained via the `ParameterizedMessage::getText(const std::string& language)` method.
- In this version of the PR, user language control is only via an explicit `userlang` query parameter. Adding support for user language selection via cookies and/or the [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header should be easy.

### CODE REVIEW

**Comment by veloman-yunkan** [comment]
@mgautierfr This PR is becoming pretty big. I think that it makes sense for you to throw a quick glance at it lest I move too far ahead with an approach that you may criticize in the end. Please note that intermediate commits contain some temporary solutions that get replaced in the end. The current state reflects my vision of how error messages and templates should be translated in the C++ code.

**Comment by kelson42** [comment]
@veloman-yunkan Thx for this PR trying to tackling the complex problem of internationalisation. Unfortunately @mgautierfr  is pretty busy these weeks and it might take a two weeks before you get a feedback from him. That said, I believe 5os PR desserves an explication comment, 5is is really too big to be obvious.

**Comment by veloman-yunkan** [comment]
@kelson42

I am planning to split this PR into at least two parts

1. Introduce unit-tests (with necessary infrastructure) that check the content of the HTML responses that are going to be internationalized. The infrastructure should support easily adding new test points for a non-English language.
2. Add internationalization. The unit tests from step 1 will help to make sure that the default output (for English) is not affected and new test points (added relying on facilities introduced in 1.) will prove that internationalization works.

File: scripts/kiwix-compile-i18n [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+# In CXX code templates, fragments that must be replaced are marked up with
+# $< and >$, while curly braces (as significant part of C++ syntax) are not
+# artificially doubled. The CXX code template is converted into a form usable
+# by Python's str.format() via 4 simple replace() operations
+def expand_cxx_template(t, **kwargs):
+    fmt = t.replace('{', '{{') \
+           .replace('}', '}}') \
+           .replace('$<', '{') \
+           .replace('>$', '}')
+    return fmt.format(**kwargs)
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply scripts/kiwix-compile-i18n:48]: I didn't know about `string.Template`. Thanks for pointing to it!

**Comment by veloman-yunkan** [comment]
After extracting more than 60% of the initial amount of changes done in the context of this PR into smaller PRs I thought that the remaining part would benefit from some history rewrite. However a quick review didn't seem to justify spending any effort on such cleanup (other than the pre-merge rebase&fixups). So I am removing the draft status from this PR.

**Comment by kelson42** [comment]
@veloman-yunkan A rebase maybe?

**Comment by veloman-yunkan** [comment]
Should we enhance the user language control with support for cookies and [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header now or postpone it until all of kiwix-serve is internationalized?

**Comment by kelson42** [comment]
> Should we enhance the user language control with support for cookies and [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header now or postpone it until all of kiwix-serve is internationalized?

AFAIK it already works (Browser language is detected and if no manual change, the language filter is set based on it), but because this is done in Javascript, the user see the change and it would be better indeed to have it in C++. That said probably in a dedicated PR... but from the timing here, sounds to be the right one.

**Comment by mgautierfr** [comment]
> Should we enhance the user language control with support for cookies and [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header now or postpone it until all of kiwix-serve is internationalized?

Yes, I was about to come back on this.
It would be better yes. We already have the Accept-Language header available. We may use it if we don't have `userlang` is not provided in the request.

I still don't know if we should have a query param, or a subpath, or .... but as @kelson42 said, it can be discussed later.

File: src/server/i18n.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+class I18nStringDB
+{
+public: // functions
+  I18nStringDB() {
+    for ( size_t i = 0; i < ARRAY_ELEMENT_COUNT(i18nStringTables); ++i ) {
+      const auto& t = i18nStringTables[i];
+      lang2TableMap[t.lang] = &t;
+    }
+		enStrings = lang2TableMap.at("en");
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/i18n.cpp:59]: Fixed

File: src/server/i18n.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    try {
+      return lang2TableMap.at(lang);
+    } catch(const std::out_of_range&) {
+      return enStrings;
+    }
+  }
+
+private: // data
+  std::map<std::string, const I18nStringTable*> lang2TableMap;
+	const I18nStringTable* enStrings;
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/i18n.cpp:84]: Fixed

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan A rebase maybe?

Done

**Comment by kelson42** [comment]
I realise the srtings file format seems not really standart. @nikkerabit Can TW bot indgest it?

**Comment by Nikerabbit** [comment]
Ping @translatewiki

**Comment by veloman-yunkan** [comment]
> > Should we enhance the user language control with support for cookies and [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header now or postpone it until all of kiwix-serve is internationalized?
> 
> Yes, I was about to come back on this. It would be better yes. We already have the Accept-Language header available. We may use it if we don't have `userlang` is not provided in the request.
> 
> I still don't know if we should have a query param, or a subpath, or .... but as @kelson42 said, it can be discussed later.

I added preliminary support for Accept-Language header. Didn't go after a language controlling cookie at this point - I think it better be addressed together with a user interface for selecting the UI language.

**Comment by veloman-yunkan** [comment]
@mgautierfr One issue with this PR (that came up relatively late in the development) is the usage of Armenian in the unit tests. When a unit-test assertion involving a multiline string containing non-ASCII characters fails, googletest (at least the version used in my build) doesn't do a good job showing the diff. Should we switch to French in the unit-tests?

**Comment by veloman-yunkan** [comment]
> I realise the srtings file format seems not really standart. @Nikerabbit Can TW bot indgest it?

@kelson42 I copied the i18n format from kiwix-desktop.

**Comment by kelson42** [comment]
@veloman-yunkan OK, then it has to be OK. No blocker on my side.
@Nikerabbit May you please secure tje strings are synced (both ways) woth TW.

**Comment by kelson42** [comment]
> @mgautierfr One issue with this PR (that came up relatively late in the development) is the usage of Armenian in the unit tests. When a unit-test assertion involving a multiline string containing non-ASCII characters fails, googletest (at least the version used in my build) doesn't do a good job showing the diff. Should we switch to French in the unit-tests?

If not already existing, would be good IMO to report or at least link this comment with an upstream bug.

**Comment by kelson42** [comment]
@mgautierfr We should probably merge this once 10.2.0 is released.
@veloman-yunkan Do you have a corresponding PR in kiwix-tools in view, so all this work becomes user visible?

**Comment by mgautierfr** [comment]
> @mgautierfr One issue with this PR (that came up relatively late in the development) is the usage of Armenian in the unit tests. When a unit-test assertion involving a multiline string containing non-ASCII characters fails, googletest (at least the version used in my build) doesn't do a good job showing the diff. Should we switch to French in the unit-tests?

On my side, if I change on test to make it fail, I have this output :

```cpp
../../SOURCE/libkiwix/test/server.cpp:950: Failure
Expected equality of these values:
  h1
    Which is: "\xD5\x8D\xD5\xAD\xD5\xA1\xD5\xAC \xD5\xB0\xD5\xA1\xD5\xBD\xD6\x81\xD5\xA5"
    As Text: "Սխալ հասցե"
  t.expectedH1
    Which is: "\xD5\x8D\xD5\xA1\xD5\xAC \xD5\xB0\xD5\xA1\xD5\xBD\xD6\x81\xD5\xA5"
    As Text: "Սալ հասցե"
Test context:
        url: /ROOT/zimfile/invalid-article
        acceptLanguageHeader: hy
```

This seems good to me.
And I like the idea to have non ascii characters in our test. It would be too easy to have bug on utf8 handling and never catch it because we only use french translation.

**Comment by mgautierfr** [comment]
> @veloman-yunkan Do you have a corresponding PR in kiwix-tools in view, so all this work becomes user visible?

We should have nothing to do on kiwix-tools side. Everything is handle "internally" by libkiwix.

**Comment by veloman-yunkan** [comment]
> > @mgautierfr One issue with this PR (that came up relatively late in the development) is the usage of Armenian in the unit tests. When a unit-test assertion involving a multiline string containing non-ASCII characters fails, googletest (at least the version used in my build) doesn't do a good job showing the diff. Should we switch to French in the unit-tests?
> 
> On my side, if I change one test to make it fail, I have this output :
> 
> ```c++
> ../../SOURCE/libkiwix/test/server.cpp:950: Failure
> Expected equality of these values:
>   h1
>     Which is: "\xD5\x8D\xD5\xAD\xD5\xA1\xD5\xAC \xD5\xB0\xD5\xA1\xD5\xBD\xD6\x81\xD5\xA5"
>     As Text: "Սխալ հասցե"
>   t.expectedH1
>     Which is: "\xD5\x8D\xD5\xA1\xD5\xAC \xD5\xB0\xD5\xA1\xD5\xBD\xD6\x81\xD5\xA5"
>     As Text: "Սալ հասցե"
> Test context:
>         url: /ROOT/zimfile/invalid-article
>         acceptLanguageHeader: hy
> ```
> 
> This seems good to me.

That's why I specifically mentioned *assertion involving a **multiline** string containing non-ASCII characters*. Try to break a test in `ServerTest.404WithBodyTesting`. The diff isn't easily comprehensible.

> And I like the idea to have non ascii characters in our test. It would be too easy to have bug on utf8 handling and never catch it because we only use french translation.

French is not ASCII-only. I think we can include a few accented characters in the tests. Their presence shouldn't spoil the diff to the same extent as when entire sentences are non-ASCII.

**Comment by veloman-yunkan** [comment]
> @mgautierfr @veloman-yunkan I don't understand what is still to be done to close [kiwix/kiwix-tools#59](https://github.com/kiwix/kiwix-tools/issues/59)?

@kelson42 There is also the welcome page (`static/skin/index.js`) that is all Javascript.

**Comment by veloman-yunkan** [comment]
> I don't have a strong opinion on this. We can merge it with Armenian test. If you prefer have french test to have more ascii friendly language, I can do a commit adding the translation and adapting the tests without problem.

Let's merge as is and set up translations by translatewiki. Then we can change the tests when (and if) they prove to cause inconvenience.

**Comment by mgautierfr** [comment]
In fact it should not be Armenian neither French. Once we put this on tranlatewiki, the translations may change relatively out of our control and break the tests. The solution is probably to have a false test lang for which we control the translation. And as it is a false translation, we can put whatever chars we want in it.

**Comment by kelson42** [comment]
.. and tests should not be translated, don't make them dependent of the strings files please.

**Comment by mgautierfr** [comment]
> .. and tests should not be translated, don't make them dependent of the strings files please.

If we want to test translation, we must check that kiwix-serve returns us translated pages. This is why I suggest we test with a fake language and be independent of real translations

**Comment by kelson42** [comment]
> This is why I suggest we test with a fake language and be independent of real translations

Yes, it is important we move on relatively quickly on this PR as if I understand right this is blocking other stuff, in particular the opensearch stuff which is urgent to merge/release. If this helps, feel free to create a ticket to be done separatly (but soon), so we can move on with this other PRs in parallel.

**Comment by mgautierfr** [comment]
Rebased-fixup on master.

I agree with the PR as it is now. I've few concern/improvement on it but it can be handle later, I will open issues for them.

**Review by mgautierfr — APPROVED** [review]

**Comment by Abijeet** [comment]
> I realise the srtings file format seems not really standart. @Nikerabbit Can TW bot indgest it?

Yes I think we should be able to ingest it. I've created [this task](https://phabricator.wikimedia.org/T306592) to track the work.

Can you confirm that the translatewiki bot will have access to commit to this repository?

**Comment by kelson42** [comment]
@Abijeet Thank you!

**Pull request merged**